lib/openshift/cluster_version.rb

```diff
@@ -1,19 +1,70 @@
 module BushSlicer
   class ClusterVersion < ClusterResource
     RESOURCE = "clusterversions"
-    def channel(user: nil, cached: true, quiet: false)
+    def channel(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('spec', 'channel')
     end
 
-    def upstream(user: nil, cached: true, quiet: false)
+    def upstream(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('spec', 'upstream')
     end
 
-    def version(user: nil, cached: true, quiet: false)
+    def version(user: nil, cached: false, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('status', 'desired', 'version')
     end
+
+    def history(user: nil, cached: false, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'history')
+    end
+
+    # return a list of history matching the state we want to filter on
+    # query_hash is the parameter we want to filter on,
+    def history_matching(query_hash: nil , user: nil, cached: true, quiet: false)
+      key = query_hash.keys.first
+      value = query_hash[key]
+      self.history.select { |h| h[key] == value }
+    end
+
+    def image(user: nil, cached: true, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'desired', 'image')
+    end
+
+    # get the last entry of the spec.history which should be where the original
+    def image_base_url(user: nil, cached: true, quiet: false)
+      self.history.last['image'].split('@sha256:')[0]
+    end
+
+    def upgrade_completed?(target_version:, user:nil, cached: false, quiet: false)
+      res = self.history_matching(query_hash: {'version' => target_version})
+      if res.count > 0
+        logger.debug("MSG: #{res.first['message']}")
+        res.first['state'] == 'Completed'
+      else
+        false
+      end
+    end
+
+    def wait_for_upgrade_completion(target_version:, user:nil, cached: false, quiet: false, upgrade_timeout: 2*60*60)
+      wait_for(upgrade_timeout) {
+        upgrade_completed?(target_version: target_version)
+      }
+    end
+    # extract the percentage done from the message field from spec.conditions
+    # default to return nil if no upgarde is being done.
+    def completed_percentage(user: nil, cached: true, quiet: false)
+      progressing_cond = self.condition(type: 'Progressing')
```

Comment:
I guess if there is no condition type `Progressing`, then any updates have finished, no?

Reply:
All of the other conditions are present, it's the extra flag that tells if it's occurring or not.  For example ```  conditions:
  - lastTransitionTime: 2019-04-04T21:45:20Z
    message: Done applying 4.0.0-0.nightly-2019-04-03-202419
    status: "True"
    type: Available
  - lastTransitionTime: 2019-04-04T21:33:16Z
    status: "False"
    type: Failing
  - lastTransitionTime: 2019-04-04T21:45:20Z
    message: Cluster version is 4.0.0-0.nightly-2019-04-03-202419
    status: "False"
    type: Progressing
  - lastTransitionTime: 2019-04-04T21:33:16Z
    message: 'Unable to retrieve available updates: unknown version 4.0.0-0.nightly-2019-04-03-202419'
    reason: RemoteFailed
    status: "False"
    type: RetrievedUpdates```

Reply:
So we get a new condition by each change such that if we upgrade again, then we will still see the above conditions + new ones?

Reply:
@akostadinov FYI, during upgrade, https://github.com/openshift-qe/output_references/blob/master/clusterversion/clusterversion_during_upgarde.yml
and after upgarde https://github.com/openshift-qe/output_references/blob/master/clusterversion/upgarde_complete.yml . The `conditions` looks like always have the same amount of element.

Reply:
but history sounds like will include more in case you do another upgrade? Is it possible to try?

Reply:
@akostadinov I'm not sure what you want me to explore here.

Reply:
Anyway, the idea for this method is to track the current upgrade percentage, we may want to track that because I've seen cases in which the number fluctuates instead of going up only.  The percentage information will only be displayed during upgrade and in the `condition` section only.